### PR TITLE
Support more types of S3 URL

### PR DESCRIPTION
**Pull request by nouse**
And set default region from EC2 metadata if using EC2 credential provider.

### CODE REVIEW

File: method/method.go [hunk 1]
```diff
@@ -124,6 +126,12 @@ func New() *Method {
 		wg:         &wg,
 		stdout:     log.New(os.Stdout, "", 0),
 	}
+	sess, _ := session.NewSession(aws.NewConfig())
```
Comment (adamvduke) [inline method/method.go]: It seems that these lines make the assumptions that:
1. The apt process is running on an instance in EC2.
1. The instance is in the same region as the s3 bucket where the apt repository is hosted.

Do I understand that correctly?

File: method/method.go [hunk 2]
```diff
@@ -229,21 +237,37 @@ func newLocation(value string) (objectLocation, error) {
 	if err != nil {
 		return objectLocation{}, nil
 	}
-	tokens := strings.Split(uri.Path, "/")
+	if uri.Host == s3Hostname {
```
Comment (adamvduke) [inline method/method.go:234]: Just to be clear, the problem being solved here is supporting bucket uri's with the form `private-repo.s3.amazonaws.com` vs `s3.amazonaws.com/private-repo`?

File: method/method.go [hunk 3]
```diff
@@ -304,9 +328,9 @@ func (m *Method) uriAcquire(msg *message.Message) {
 // s3Client provides an initialized s3iface.S3API based on the contents of the
 // provided url.URL. The access key id and secret access key are assumed to
 // correspond to the Username() and Password() functions on the URL's User.
-func (m *Method) s3Client(s3Uri *url.URL) s3iface.S3API {
-	awsAccessKeyID := s3Uri.User.Username()
-	awsSecretAccessKey, hasPass := s3Uri.User.Password()
+func (m *Method) s3Client(user *url.Userinfo) s3iface.S3API {
```
Comment (adamvduke) [inline method/method.go:325]: 👍

File: method/method.go (hunk 1 above)
Comment (nouse) [reply method/method.go]: Yes, that will save effort to set region in apt config when bucket is same region with instance.

File: method/method.go (hunk 2 above)
Comment (nouse) [reply method/method.go:234]: Yes, now we support 3 types of s3 repo URL, s3://<bucket>, s3://<bucket>.s3.amazonaws.com besides current one.

**Review by adamvduke — COMMENTED** [review]
I like the change that adds support for additional repo URLs. Do you mind moving just that change to its own pull request?

I would prefer not to merge the change that sets the region based on `ec2metadata` because of the assumptions about the process running on ec2 and the bucket being in the same region. Also, that seems to be a breaking change for ec2 instances running outside of us-east-1 where the bucket is in us-east-1.

**Comment by nouse** [comment]
I want to add reading region from S3 URL, https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro. What's your idea?

I will put it in different PR.

**Comment by nouse** [comment]
> I like the change that adds support for additional repo URLs. Do you mind moving just that change to its own pull request?
> 
> I would prefer not to merge the change that sets the region based on `ec2metadata` because of the assumptions about the process running on ec2 and the bucket being in the same region. Also, that seems to be a breaking change for ec2 instances running outside of us-east-1 where the bucket is in us-east-1.

Done.

**Review by adamvduke — APPROVED** [review]
Looks great. Thanks for contributing.

**Comment by adamvduke** [comment]
> I want to add reading region from S3 URL, https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro. What's your idea?
> 
> I will put it in different PR.

I'm not sure I have a great solution for setting the region from the URL. Apt has a well defined scheme for setting configuration options through configuration files and properties that are passed into the code that downloads files.

**Pull request merged**